src/parser/jobs/rdm/Procs.js

```diff
@@ -0,0 +1,322 @@
+import React, {Fragment} from 'react'
+import {ActionLink, StatusLink} from 'components/ui/DbLink'
+import ACTIONS from 'data/ACTIONS'
+import STATUSES from 'data/STATUSES'
+import Module from 'parser/core/Module'
+import {TieredSuggestion} from 'parser/core/modules/Suggestions'
+import {CAST_STATE, PROCS, SEVERITY_MISSED_PROCS, SEVERITY_OVERWRITTEN_PROCS, SEVERITY_INVULN_PROCS} from 'parser/jobs/rdm/ProcsEnum'
+import {i18nMark} from '@lingui/react'
+
+export default class Procs extends Module {
+	static handle = 'procs'
+	static title = 'Procs'
+	static dependencies = [
+		'aoe',
+		'castTime',
+		'downtime',
+		'gcd',
+		'suggestions',
+		'gauge',
+	]
+	static i18n_id = i18nMark('rdm.procs.title')
+
+	_history = {
+		wastedStone: 0,
+		overWrittenStone: 0,
+		invulnStone: 0,
+		missedStone: 0,
+		wastedFire: 0,
+		overWrittenFire: 0,
+		invulnFire: 0,
+		missedFire: 0,
+		wastedImpactful: 0,
+		overWrittenImpactful: 0,
+		invulnImpactful: 0,
+		missedImpactful: 0,
+		wastedEnhancedScatters: 0,
+		invulnScatter: 0,
+		missedScatter: 0,
+	}
+	_currentProcs = {
+		verStoneGained: 0,
+		verFireGained: 0,
+		impactfulGained: 0,
+		enhancedScatterGained: 0,
+	}
+	_castState = CAST_STATE.NA
+	_impactfulProcOverride = false
+	_previousCast = null
+	_bossWasInvuln = false
+
+	constructor(...args) {
+		super(...args)
+
+		this.addHook('cast', {by: 'player'}, this._onCast)
+		this.addHook('applybuff', {
+			to: 'player',
+			abilityId: PROCS,
+		}, this._onGain)
+		this.addHook('removebuff', {
+			to: 'player',
+			abilityId: PROCS,
+		}, this._onRemove)
+		this.addHook('complete', this._onComplete)
+	}
+
+	_onCast(event) {
+		const abilityID = event.ability.guid
+		const invuln = this.downtime.getDowntime(this._previousCast||0, event.timestamp)
+		this._previousCast = event.timestamp
+		this._bossWasInvuln = invuln > 0
+
+		switch (abilityID) {
+		case ACTIONS.IMPACT.id:
+			this._castState = CAST_STATE.IMPACT
+			const impactRemainingDuration = event.timestamp - this._currentProcs.impactfulGained
+			if (impactRemainingDuration <= 8000) {
+				//Use Impactful at 8 or less seconds remaining, regardless of all other procs
+				this._impactfulProcOverride = true
+			}
+			break
+		case ACTIONS.VERSTONE.id:
+			this._castState = CAST_STATE.VERSTONE
+			break
+		case ACTIONS.VERFIRE.id:
+			this._castState = CAST_STATE.VERFIRE
+			break
+		case ACTIONS.VERTHUNDER.id:
+			this._castState = CAST_STATE.VERTHUNDER
+			break
+		case ACTIONS.VERAREO.id:
+			this._castState = CAST_STATE.VERAREO
+			break
+		case ACTIONS.SCATTER.id:
+			this._castState = CAST_STATE.SCATTER
+			break
+		default:
+			this._castState = CAST_STATE.NA
+			break
+		}
+	}
+
+	_onGain(event) {
+		const statusID = event.ability.guid
+
+		switch (statusID) {
+		case STATUSES.VERSTONE_READY.id:
```

Comment:
This could likely be simplified by instead having a map from STATUS -> some counter object combining a lot of this, which would generalize this entire switch statement.

Reply:
If I understand you correctly soemthing like below?  Or are you thinking something more akin to property iteration of this._currentProcs with some mapping to know which overWritten property to update?  Or as a tertiary option instead of just a list of counters are you wanting more of an array of occurrences with the type of occurrence (Overwrite, Missed, etc) and the StatusID so I know which spell it belonged to?

STATUSES.map(ID => {
//Just copy the if statements current in cases and string them here, removing the switch/cast lines?
})

Reply:
Something along the lines of

```javascript
const _missedProcs = {};
[...]

const missed = _currentProcs[statusId] > 0 && !this._impactfulProcOverride
_missedProcs[statusId] = (_missedProcs[statusId] || 0) + (missed ? 1 : 0)
```